src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ConnectionPool/HttpConnectionPool.Http3.cs

```diff
@@ -100,103 +114,523 @@ internal sealed partial class HttpConnectionPool
+        private async Task InjectNewHttp3ConnectionAsync(RequestQueue<Http3Connection?>.QueueItem queueItem)
+        {
+            Debug.Assert(IsHttp3Supported());
+
+            if (NetEventSource.Log.IsEnabled()) Trace("Creating new HTTP/3 connection for pool.");
+
+            // Queue the remainder of the work so that this method completes quickly
+            // and escapes locks held by the caller.
+            await Task.CompletedTask.ConfigureAwait(ConfigureAwaitOptions.ForceYielding);
+
+            Http3Connection? connection = null;
+            Exception? connectionException = null;
+            HttpAuthority? authority = null;
+            HttpConnectionWaiter<Http3Connection?> waiter = queueItem.Waiter;
+
+            CancellationTokenSource cts = GetConnectTimeoutCancellationTokenSource();
+            waiter.ConnectionCancellationTokenSource = cts;
+            try
+            {
+                if (TryGetHttp3Authority(queueItem.Request, out authority, out Exception? reasonException))
                 {
-                    // Connection expired.
-                    if (NetEventSource.Log.IsEnabled()) http3Connection.Trace("Found expired HTTP3 connection.");
-                    http3Connection.Dispose();
-                    InvalidateHttp3Connection(http3Connection);
+                    // If the authority was sent as an option through alt-svc then include alt-used header.
+                    connection = new Http3Connection(this, authority, includeAltUsedHeader: _http3Authority == authority);
+
+                    QuicConnection quicConnection = await ConnectHelper.ConnectQuicAsync(queueItem.Request, new DnsEndPoint(authority.IdnHost, authority.Port), _poolManager.Settings._pooledConnectionIdleTimeout, _sslOptionsHttp3!, connection.StreamsAvailableCallback, cts.Token).ConfigureAwait(false);
+                    if (quicConnection.NegotiatedApplicationProtocol != SslApplicationProtocol.Http3)
+                    {
+                        await quicConnection.DisposeAsync().ConfigureAwait(false);
+                        throw new HttpRequestException(HttpRequestError.ConnectionError, "QUIC connected but no HTTP/3 indicated via ALPN.", null, RequestRetryType.RetryOnConnectionFailure);
+                    }
+                    connection.InitQuicConnection(quicConnection);
                 }
-                else
+                else if (reasonException is not null)
+                {
+                    ThrowGetVersionException(queueItem.Request, 3, reasonException);
+                }
+            }
+            catch (Exception e)
+            {
+                connectionException = e is OperationCanceledException oce && oce.CancellationToken == cts.Token && !waiter.CancelledByOriginatingRequestCompletion ?
+                    CreateConnectTimeoutException(oce) :
+                    e;
+                // If the connection hasn't been initialized with QuicConnection there's no need to dispose it.
```

Comment:
This assumes that `InitQuicConnection` may never throw, right?

Reply:
Yes, but it got me thinking and it looks like, we don't need this in the end as the H/3Connection Dispose does check for `_connection` not being null, so calling `Dispose` on not fully initialized connection should not do any harm.